packages/langium/src/utils/grammar-utils.ts

```diff
@@ -234,29 +234,41 @@ export function findAssignment(cstNode: CstNode): ast.Assignment | undefined {
  * from a parser rule, and that rule must contain an assignment to the `name` property. In all other cases,
  * this function returns `undefined`.
  */
-export function findNameAssignment(type: ast.AbstractType | ast.InferredType): ast.Assignment | undefined {
-    if (ast.isInferredType(type)) {
-        // inferred type is unexpected, extract AbstractType first
-        type = type.$container;
+export function findNameAssignment(type: ast.AbstractType): ast.Assignment | undefined {
+    let startNode: AstNode = type;
+    if (ast.isInferredType(startNode)) {
+        // for inferred types, the location to start searching for the name-assignment is different
+        if (ast.isAction(startNode.$container)) {
+            // a type which is explicitly inferred by an action: investigate the sibbling of the Action node, i.e. start searching at the Action's parent
+            startNode = startNode.$container.$container!;
+        } else if (ast.isParserRule(startNode.$container)) {
+            // investigate the parser rule with the explicitly inferred type
+            startNode = startNode.$container;
+        } else {
+            throw new Error('impossible');
```

Comment:
May use assertUnreachable?

Reply:
Yes, that is much better, thanks for the hint!